File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <stdlib.h>
#include <math.h>

dnd_character_t make_dnd_character() {
    dnd_character_t character; 
    character.strength = ability();
    character.dexterity = ability();
    character.constitution = ability();
    character.intelligence = ability();
    character.wisdom = ability();
    character.charisma = ability();

    character.hitpoints = 10 + modifier(character.constitution);
    
    return character;
}

int ability() {
    int sum = 0;
    int min_index = 0;
    
    int upper_bound = 6; 
    int lower_bound = 1; 

    int rolls[4];

    // roll four die
    for (int i = 0; i < 4; i++) { 
        int roll = rand() % (upper_bound - lower_bound + 1) + lower_bound; 
        rolls[i] = roll;
    } 

    // find min index
    for (int i = 0; i < 4; ++i) {
        if (rolls[i] < rolls[min_index]) {
            min_index = i;
        }
    }

    // sum other three rolls
    for (int i = 0; i < 4; ++i) {
        if(i != min_index) {
            sum += rolls[i];
        }
    }
    return sum;
}

int modifier(int score) {
    return floor((score - 10) / 2.0);
}

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

class DnDCharacter {

    int strength = ability(rollDice());
    int wisdom = ability(rollDice());
    int intelligence = ability(rollDice());
    int charisma = ability(rollDice());
    int dexterity = ability(rollDice());
    int constitution = ability(rollDice());

    int ability(List<Integer> scores) {

        // find min
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < scores.size(); i++) {
            if (scores.get(i) < min) {
                min = scores.get(i);
            }
        }
        return scores.get(0) + scores.get(1) + scores.get(2) + scores.get(3) - min;
    }

    List<Integer> rollDice() {
        List<Integer> scores = new ArrayList<>();
        Random rand = new Random();
        for (int i = 0; i < 4; i++) {
            int score = rand.nextInt(6) + 1;
            scores.add(score);
        }
        return scores;
    }

    int modifier(int input) {
        return (int) Math.floor((input - 10) / 2.0);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return 10 + modifier(getConstitution());
    }
}
